### DISCUSSION TITLE
How to dynamically change window position?

### DISCUSSION THREAD

**Discussion by 9mm** [macvim-dev/macvim#1398] [Q&A]
Hey there, I have the following options set:

* `Hidden title bar`
* `Smoothly resizes window`

The problem is when I use a window manager, it looks like this. This is with 2 windows set 50/50 (notice the missing title bar is throwing off the calculation)

<img width="907" alt="image" src="https://user-images.githubusercontent.com/1386207/227267480-72f7a8ca-9aff-48c7-bac6-7ce98fe72c63.png">

Is there a way I can fix this? I'm assuming I can't fix the actual problem, but is there a way I can use the window resize function somehow to do it on startup?

This is the code I'm using in my `.vimrc`, but I notice it is sometimes very buggy and 'random', and I'm not quite sure why... is there a way I can make the following code work deterministically?

```
" MacVim
if has('gui_running')
  set window location to align nicely on start
  autocmd VimEnter * call timer_start(120, { tid -> execute('winpos 893 0')})
endif
```

**Comment by 9mm** [comment]
Hmm so actually, on my new M2 silicon mac the window is much more accurate... now my question is, instead of resizing by columns (which doesnt have pixel perfect precision), given that I have 'smooth resize' on is there a way I can change per-pixel width of the macvim window? (instead of using `win`)

```
  " set window location to align nicely on start
  " 105 / 870 or 106 / 865
  execute('win 106 64')
  autocmd VimEnter * call timer_start(120, { tid -> execute('winpos 865 0')})
```

**Reply by ychin** [reply]
Yeah I'm also not seeing an issue with the winpos as well as it seems to work fine. What's the OS / machine type etc / MacVim version etc for the one that doesn't work well?

> now my question is, instead of resizing by columns (which doesnt have pixel perfect precision), given that I have 'smooth resize' on is there a way I can change per-pixel width of the macvim window? (instead of using win)

The answer is there isn't one right now. You can use `macaction zoomRight:` to pin to the right half, or `macaction performZoom:` to maximize the window, but other than that there is no pixel-accurate way to control window size.

Is there a reason you need such accuracy? Have you thought about using a dedicated tool like BetterSnapTool etc?

**Reply by 9mm** [reply]
I need the accuracy because I'm trying to get my windows to line up manually, simply because those window managers do not work when the menu bar is hidden.

For some reason size-up did it perfectly, but "rectangle" and some others I have tried do not. Sizeup isnt maintained anymore though so I was going to switch.

Heres the image again of what happens when using window managers/ better touch tool etc

![image](https://user-images.githubusercontent.com/1386207/227367751-dc375896-32db-4df3-ab22-483f5c3dc1ef.png)

**Reply by ychin** [reply]
Ah ok I think I didn't understand what you meant in the original post by the gap caused by the window managers.

You can always go the old-school Apple way of using AppleScript:

```
:!osascript -e 'tell application "MacVim"' -e 'set bounds of front window to {10, 20, 640, 480}' -e 'end tell'
```

Because MacVim is controlling itself I don't believe it will need to prompt for permissions. Otherwise if you are calling this from the terminal it may prompt you for that.

**Comment by eirnym** [comment]
Which version of MacVim do you use? 
Which macOS do you use?
How do you install MacVim? 

Is this problem reproducible with `mvim --clean`?

**Comment by eirnym** [comment]
I set similar preferences on my MacVim and located Apple Terminal and MacVim side by side. 

The `winpos` command shows X 735 Y 37, and even I explicitly set Y to 0 using it, it still shows 37. 

After moving window around I repeated the command with same parameters and it landed nicely. 

My macOS is Monterey on M2, downloaded from GitHub releases. 

Note: I haven't tried your snippet

**Reply by ychin** [reply]
> and even I explicitly set Y to 0 using it, it still shows 37.

That's because 0 is the top of the screen occupied by the menu, and the OS won't let you nudge it that high.
